packages/streamr-eventemitter/include/streamr-eventemitter/EventEmitter.hpp

```diff
@@ -0,0 +1,190 @@
+
+    public:
+        static HandlerReference create() {
+            // Use a "magic static"
+            // https://blog.mbedded.ninja/programming/languages/c-plus-plus/magic-statics/
+            static size_t counter = 1;
+            return HandlerReference(counter++);
+        }
+        static HandlerReference createNonExistent() {
+            return HandlerReference(0);
+        }
+        [[nodiscard]] size_t getId() const { return id; }
+    };
+
+public:
+    template <typename EventType, typename CallbackType>
+        requires(
+            std::is_same<EventType, EmitterEventType>::value &&
+            std::is_assignable<
+                typename EmitterEventType::Handler::HandlerFunction,
+                CallbackType>::value)
+    HandlerReference on(const CallbackType& callback, bool once = false) {
+        typename EmitterEventType::Handler::HandlerFunction handlerFunction =
+            callback;
+        // silently ignore null callbacks
+        if (!handlerFunction) {
+            return HandlerReference::createNonExistent();
+        }
+        auto handlerReference = HandlerReference::create();
+        typename EmitterEventType::Handler handler(
+            handlerFunction, handlerReference.getId(), once);
+
+        std::lock_guard guard{mutex};
+
+        this->eventHandlers.push_back(handler);
+        return handlerReference;
+    }
+
+    template <typename EventType, typename CallbackType>
+        requires(
+            std::is_same<EventType, EmitterEventType>::value &&
+            std::is_assignable<
+                typename EmitterEventType::Handler::HandlerFunction,
+                CallbackType>::value)
+    HandlerReference once(const CallbackType& callback) {
+        return this->on<EventType, CallbackType>(callback, true);
+    }
+
+    template <typename EventType>
+        requires(std::is_same<EventType, EmitterEventType>::value)
+    void off(HandlerReference handlerReference) {
+        std::lock_guard guard{mutex};
+
+        this->eventHandlers.remove_if(
+            [&handlerReference](
+                const typename EmitterEventType::Handler& handler) {
+                return handler.getId() == handlerReference.getId();
+            });
+    }
+
+    template <typename EventType>
+        requires(std::is_same<EventType, EmitterEventType>::value)
+    uint32_t listenerCount() {
+        std::lock_guard guard{mutex};
+        return this->eventHandlers.size();
+    }
+
+    template <typename EventType>
+        requires(std::is_same<EventType, EmitterEventType>::value)
+    void removeAllListeners() {
+        std::lock_guard guard{mutex};
+        this->eventHandlers.clear();
+    }
+
+    template <typename EventType, typename... EventArgs>
+        requires(std::is_same<EventType, EmitterEventType>::value)
+    void emit(EventArgs&&... args) {
+        std::list<typename EmitterEventType::Handler> currentHandlers;
+
+        { // Take copy of eventHandlers at the time of emitting
```

Comment:
I don't understand why the copy of event handlers is really needed.

Reply:
It is done in order to emit the event to only those handlers that were there when the emit() call was made. We need to follow the rule "If emitter E has listeners L and K when emit() gets called, event gets propagated exactly to listeners L and K" If we don't make the copy, listeners may appear or be removed during emitting process, and the rule does not hold anymore.

Reply:
It might even be necessary to prevent the calling of on(), off() and once() with a mutex during the emitting loop . The current implementation that copies the listeners might have the problem that if off(L) gets called during the execution of the loop, and the caller of off(L) deletes L, we have an invalid reference to L. 

On the other hand, executing the event loop might (and in our use case, probably will) take a long time. During this time on(), off() and once() would be blocked.